src/NHibernate.Test/Criteria/Lambda/ExpressionProcessorFindValueFixture.cs

```diff
@@ -42,6 +42,36 @@ public void StaticPropertyInstanceMethodCall()
 
 			Assert.AreEqual(expected, actual);
 		}
+		
+		[Test]
+		public void NullableNullHasValue()
+		{
+			int? v = null;
+			var actual = GetValue(() => v.HasValue);
+			var expected = v.HasValue;
+
+			Assert.AreEqual(expected, actual);
+		}		
+		
+		[Test]
+		public void NullableNullGetValueOrDefault()
+		{
+			int? v = null;
+			var actual = GetValue(() => v.GetValueOrDefault());
+			var expected = v.GetValueOrDefault();
+
+			Assert.AreEqual(expected, actual);
+		}
+		
+		[Test]
+		public void NullableNullValue()
+		{			
+			int? v = null;
+			Expression<Func<int>> expression = () => v.Value;
+
+			Assert.Throws<InvalidCastException>(() => GetValue(expression));
+			Assert.Throws<InvalidCastException>(() => expression.Compile().Invoke());
+		}
```

Comment:
What about `e => v.HasValue || e.SomeProperty = v.Value`? I wonder if it supported or if it would throw.

(There is not much value in supporting this, excepted "feature parity" with linq-to-object: the `.Value` is not needed in any way in such a query and can just be dropped. But still, it is better to know what how it behaves with NHibernate.)

Reply:
It throws on `v.Value` evaluation:
```C#
    NHibernate.HibernateException : Evaluation failure on null.Value
  ----> System.InvalidOperationException : Nullable object must have a value.
```
Maybe we shouldn't try to evaluate it in this context. But it's a different story not related to this PR.

Reply:
For sure, it would be beneficial if `nullable.Value` works (with the SQL parameter becoming false, true, or NULL). The value of supporting this would be that users don't have to remember to adapt their expression specially.

No idea what the effort would be to support it though.

Reply:
> it would be beneficial if `nullable.Value` works

Just to clarify things. This PR fixes regression caused by optimizations made in https://github.com/nhibernate/nhibernate-core/pull/2948, https://github.com/nhibernate/nhibernate-core/pull/2957. `nullableNullParam.HasValue` or `nullableNullParam.GetValueOrDefault()` works in 5.3 but throws on master (5.4-dev).

`nullable.Value` fails the same in 5.3. So it's more like another feature request.